agent/job_runner.go

```diff
@@ -386,6 +388,29 @@ func (r *JobRunner) normalizeVerificationBehavior(behavior string) (string, erro
 	}
 }
 
+// filterEnv removes the keys from the map that don't exist in the slice.
+func filterEnv(env map[string]string, allowedVariables []string) []string {
+	// Filter out if the user has unlisted environment variables
+	// It adds environment variable to denied list and removes if it finds matching regex
+	var deniedVariables []string
+	for k := range env {
+		deniedVariables = append(deniedVariables, k)
+		for _, v := range allowedVariables {
+			if isMatch, _ := regexp.MatchString(v, k); isMatch {
+				deniedVariables = deniedVariables[:len(deniedVariables)-1]
+				break
+			}
+		}
+	}
+
+	// Delete denied variables from environment variables list
+	for _, v := range deniedVariables {
+		delete(env, v)
+	}
+
+	return deniedVariables
+}
+
```

Comment:
this approach of stripping disallowed environment variables out is quite different to the ones in https://github.com/buildkite/agent/pull/2361 and https://github.com/buildkite/agent/pull/2394 - both of them refuse the job, instead of stripping the disallowed values out. is there a reason that you decided to do this?

Reply:
I suggested this approach rather than keeping a boolean that is checked after iterating finishes in L#398, but I do see your point in the consistency of building the set.

Reply:
Ah, I think @moskyb meant why it's not rejecting the job if it has bad variables set, unlike other PRs.

And that's actually really a good point. The reason for just filtering out is that an unknown repo or plugin can actually run some malicious code. However, with environment variables, I thought it's quite limited what attackers can do with it. So I just made it to filter the unknown values.